src/rpc/blockchain.cpp

```diff
@@ -2268,17 +2269,49 @@ class BlockFiltersScanReserver
     }
 };
 
+static bool CheckBlockFilterMatches(BlockManager& blockman, const CBlockIndex& blockindex, const GCSFilter::ElementSet& needles)
+{
+
+    const CBlock block{GetBlockChecked(blockman, &blockindex)};
+    const CBlockUndo block_undo{GetUndoChecked(blockman, &blockindex)};
+
+    for (const auto& needle : needles) {
+        const auto script{CScript(needle.begin(), needle.end())};
+
+        // Check if any of the outputs match the scriptPubKey
+        for (const auto& tx : block.vtx) {
+            if (std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& txout) { return txout.scriptPubKey == script; })) {
+                return true;
+            }
+        }
+
+        // Check if any of the inputs match the scriptPubKey
+        for (const auto& txundo : block_undo.vtxundo) {
+            if (std::any_of(txundo.vprevout.cbegin(), txundo.vprevout.cend(), [&](const auto& coin) { return coin.out.scriptPubKey == script; })) {
+                return true;
+            }
+        }
+    }
```

Comment:
ISTM this outer loop isn't necessary and could potentially be quite slow. `GCSFilter::ElementSet` is a `std::unordered_set` which has constant time lookup. But with this loop, the lookup becomes linear. It should be much faster to see if each scriptPubKey is contained in `needles`.

```suggestion
// Check if any of the outputs match the scriptPubKey
    for (const auto& tx : block.vtx) {
        if (std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& txout) { return needles.count(std::vector<unsigned char>(txout.scriptPubKey.begin(), txout.scriptPubKey.end())) != 0; })) {
            return true;
        }
    }

    // Check if any of the inputs match the scriptPubKey
    for (const auto& txundo : block_undo.vtxundo) {
        if (std::any_of(txundo.vprevout.cbegin(), txundo.vprevout.cend(), [&](const auto& coin) { return needles.count(std::vector<unsigned char>(coin.out.scriptPubKey.begin(), coin.out.scriptPubKey.end())) != 0; })) {
            return true;
        }
    }
```